### ISSUE TITLE
Usage with ISR (Nextjs 13)

### ISSUE THREAD

**Issue by arduano** [sanity-io/next-sanity#177]
In NextJS 13, they changed the way ISR is handled, with extra props passed into the `fetch` call rather than any external config. See https://nextjs.org/blog/next-13#data-fetching.

Is it possible to achieve SSR or ISG in Next 13 this using this library?

**Comment by arduano** [comment]
So I have found that you can `export const revalidate = <seconds>` in the page, which didn't appear in any of NextJS's blog posts that I've seen, that mostly solves my issue. However, I still haven't found how to configure SSR rather than ISR so I'll keep this issue open.

**Comment by stipsan** [comment]
Hi @arduano, we're exploring this [exact use case in our new starter](https://github.com/sanity-io/nextjs-blog-cms-sanity-v3/pull/90) :relaxed:

**Comment by wimbarelds** [comment]
To double-check, is the plan/intention for `SanityClient.fetch` to have some means of passing an argument along the lines of `{ next: { revalidate: 60 }`?

**Comment by stipsan** [comment]
> To double-check, is the plan/intention for `SanityClient.fetch` to have some means of passing an argument along the lines of `{ next: { revalidate: 60 }`?

We're exploring ways to support revalidation, yes. But it's not decided what that'll look like.

**Comment by Blankeos** [comment]
> So I have found that you can `export const revalidate = <seconds>` in the page, which didn't appear in any of NextJS's blog posts that I've seen, that mostly solves my issue. However, I still haven't found how to configure SSR rather than ISR so I'll keep this issue open.

The snippet you showed definitely works. I personally found that in this [tutorial](https://www.youtube.com/watch?v=x3fCEPFgUSM). That said, I'm still looking for a way using SSR with the new data fetching conventions on Next 13. Has anyone found a way?

If not, maybe a way to disable caching because my homepage doesn't exactly update the new data.

**Comment by ringods** [comment]
The App Router is now marked `Stable` in [Next.js 13.4](https://nextjs.org/blog/next-13-4). Is there any update on support for the extended `fetch` call in this package?

**Comment by markomitranic** [comment]
Hey @stipsan are there any news on this front? The (not so)new Next App Router is pretty much the default and the future of Next at this point, as it's been around for a year.

I've been using [this library](https://github.com/fedeya/next-sanity-client) as a client, which isn't a great experience, and we'd rather see this solved within Sanity client itself, instead of having to wrap it ourselves. Using `const revalidate` is not a permanent solution, as it circumvents with Next's most powerful approuter feature.

**Comment by stipsan** [comment]
> Hey @stipsan are there any news on this front? The (not so)new Next App Router is pretty much the default and the future of Next at this point, as it's been around for a year.

We've started planning an upgrade to our tool chain that will use `fetch` internally, consistently (instead of the NodeJS request libraries), and that will allow us to fully support the new features like setting cache, revalidate and more, on a data fetching level. That's all I can share right now 🙌

**Comment by JayBox325** [comment]
Really been struggling with revalidation. Is there any way to stick with this package and make it work? I looked into swapping to the package @markomitranic suggested, but I didn't like moving away from `next-sanity`.

**Comment by stipsan** [comment]
@JayBox325 we have experimental and undocumented support:
```bash
npm i @sanity/client@latest next-sanity@latest
```

```tsx
import {createClient} from 'next-sanity'

const client = createClient({
  // standard config options
})

export async function ServerComponent() {
  const data = await client.fetch(`*[_type =="page"]`, params, {
    // These options are new
    cache: 'force-cache',
    next: { revalidate: 60 * 60, tags: ['pages'] }
  })
}
```

We're currently doing QA and end-to-end testing on these features. Once that completes we'll publish documentation, update relevant README's, update our templates to use it, and notify this thread 🙌

**Comment by JayBox325** [comment]
This is great to know, however, throws this message when you include both:
```
specified "cache: force-cache" and "revalidate: 60", only one should be specified.
```
So that might be handy to include in the docs.

The issue I'm facing with Next now is this AbortError being returned with every request and some content staying stale:
```
DOMException [AbortError]: This operation was aborted
    at Object.fetch (node:internal/deps/undici/undici:11457:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```
I'm struggling to identify if it's a Next 13.4.13 or next-sanity thing.

**Comment by stipsan** [comment]
> This is great to know, however, throws this message when you include both:
> ```specified "cache: force-cache" and "revalidate: 60", only one should be specified.```
> So that might be handy to include in the docs.

Thanks, snippet updated! The snippet was showing all the options available, in real world usage we typically have a `sanity.fetch.ts` that looks like this:
```tsx
import 'server-only'

import { createClient, type QueryParams } from '@sanity/client'
import { draftMode } from 'next/headers'

const client = createClient({
  projectId: process.env.NEXT_PUBLIC_SANITY_PROJECT_ID || 'pv8y60vp',
  dataset: process.env.NEXT_PUBLIC_SANITY_DATASET || 'production',
  apiVersion: process.env.NEXT_PUBLIC_SANITY_API_VERSION || '2022-11-15',
  useCdn: false,
  perspective: 'published',
})

// Used with `LiveQueryProvider`
export const token = process.env.SANITY_API_READ_TOKEN!

const DEFAULT_PARAMS = {} as QueryParams
const DEFAULT_TAGS = [] as string[]

export async function sanityFetch<QueryResponse>({
  query,
  params = DEFAULT_PARAMS,
  cache = 'force-cache',
  tags = DEFAULT_TAGS,
}: {
  query: string
  params?: QueryParams
  cache?: RequestCache
  tags: string[]
}): Promise<QueryResponse> {
  const isDraftMode = draftMode().isEnabled
  if (isDraftMode && !token) {
    throw new Error(
      'The `SANITY_API_READ_TOKEN` environment variable is required.',
    )
  }

  return client.fetch<QueryResponse>(query, params, {
    cache,
    ...(isDraftMode && {
      token,
      perspective: 'previewDrafts',
    }),
    next: {
      ...(isDraftMode && { revalidate: 0 }),
      tags,
    },
  })
}

```

> The issue I'm facing with Next now is this AbortError being returned with every request and some content staying stale

This one could be on our side but I would need a repro (CodeSandbox link would be amazing) to investigate. We haven't seen errors like these in our own testing :think

**Comment by JayBox325** [comment]
Yours is a lot more comprehensive than ours (as we don't have live preview enabled yet), but it's great to see so thanks for sharing! I may be happy to make a repro of my actual live working project if you want? Or at least a copy of, would that be helpful?

**Comment by stipsan** [comment]
Yeah that would be super helpful @JayBox325! 😄

**Comment by JayBox325** [comment]
Ok, I'll sort you out as a collaborator on an isolated reproduction. If you need to talk anything through let me know in Slack and we'll get this sorted.

**Comment by stipsan** [comment]
@JayBox325 feel free to reach out to me over slack and we can take it from there 👍

**Comment by markomitranic** [comment]
@stipsan @JayBox325 thank you for all the effort :) please update this issue if you get any new feedback as it may end up being super useful to the rest of us!

**Comment by stipsan** [comment]
The AbortError case:
```
DOMException [AbortError]: This operation was aborted
    at Object.fetch (node:internal/deps/undici/undici:11457:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```
Seems to be resolved in [`@sanity/client` v6.4.7](https://github.com/sanity-io/client/releases/tag/v6.4.7)

**Comment by stipsan** [comment]
Closing this as we now have ISR implemented in our [Personal website template for App Router](https://github.com/sanity-io/sanity-template-nextjs-app-router-personal-website). We've also updated the README with App Router specific information on how to setup a webhook handler that calls `revalidateTag`, as well as how to pass `next.tags` into a `client.fetch` call.

If you have any questions or need help setting up, feel free to open a new issue or reach out on the #nextjs channel on our https://slack.sanity.io
